### PR TITLE
Update README

### PR DESCRIPTION
**Pull request by jacob-hughes**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -123,6 +130,43 @@ the official Rust compiler:
...
+}
+```
+
+This creates a garbage collected object containing the `u64` value `123`. A
+`Gc`'s data is stored in a `GcBox` internally (similar to `Rc`'s
+implementation).
+
+### Interior mutability
+
+There is no way to mutate, or obtain a mutable reference (`&mut T`) to a `Gc<T>`
```
Comment (ltratt) [inline README.md:153]: s/to a/to the contents of a/ ?

File: README.md [hunk 2]
```diff
@@ -144,6 +188,28 @@ parallel (but not concurrent!)[^1] collection.
...
+Finalisers are a common component of most tracing GCs which are used to run code
+for cleanup once an object dies (e.g.~closing a file handle or a database
+connection).
+
+Alloy takes a novel approach to finalisation compared to previous Rust GCs in
+that it uses existing drop methods as garbage collection finalisers, saving
+users the potentially error-prone task of manually writing both destructor and
+finaliser methods for GC managed objects.
+
+When a `Gc`'s underlying `GcBox`  becomes unreachable, Alloy will call its
```
Comment (ltratt) [inline README.md:202]: s/underlying GcBox/underlying contents/ ?

File: README.md [hunk 3]
```diff
@@ -123,6 +130,43 @@ the official Rust compiler:
...
+
+```rust
+use std::gc::Gc;
+
+fn main() {
+    let a = Gc::new(123);
+}
+```
+
+This creates a garbage collected object containing the `u64` value `123`. A
```
[the remark this answers is not in the document]
Comment (jacob-hughes) [reply README.md]: You're right, I've removed it: 76e8b2eea8d6452fdcdd7114db6ce30c0a9806f4

File: README.md (hunk 1 above)
Comment (jacob-hughes) [reply README.md]: Fixed here: 01bf7083554ca91768c138e4b4fe66d9448a3d13

File: README.md [hunk 4]
```diff
@@ -144,6 +188,28 @@ parallel (but not concurrent!)[^1] collection.
     collector simply means that the garbage collection workload can be
     parallelised across multiple worker threads.
 
+### Finalisation
+
+Finalisers are a common component of most tracing GCs which are used to run code
+for cleanup once an object dies (e.g.~closing a file handle or a database
+connection).
+
+Alloy takes a novel approach to finalisation compared to previous Rust GCs in
```
[the remark this answers is not in the document]
Comment (jacob-hughes) [reply README.md:195]: I've reworked this section a bit. What do you think?

File: README.md (hunk 2 above)
Comment (jacob-hughes) [reply README.md]: Fixed in the rework of the finaliser section.

File: README.md [hunk 5]
```diff
@@ -199,16 +199,54 @@ that it uses existing drop methods as garbage collection finalisers, saving
...
+12 |     let x = Gc::new(Node { next: Gc::new(123) });
+   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^ has a drop method which cannot be safely finalized.
+...
+2  |         *self.next;
+   |          ---------
+   |          |
+   |          caused by the expression here in `fn drop(&mut)` because
+   |          it uses another `Gc` type.
+   |
+   = help: `Gc` finalizers are unordered, so this field may have already been dropped. It is not safe to dereferenc
```
Comment (ltratt) [inline README.md:246]: Unfortunate line brea

k.

File: README.md [hunk 6]
```diff
@@ -199,16 +199,54 @@ that it uses existing drop methods as garbage collection finalisers, saving
...
 finaliser, which means that `drop` is called on all the component types (in the
 same way that Rust automatically calls `drop` in an RAII context).
 
-#### TODO: still need to explain
+To achieve Alloy's goal of making cyclic data structures easier to write, we
+made the decision to support the finalisation of objects with cycles. This means
+that no guarantees are made about the order in which finalisers are run. In
+order for this to be sound, you must not access other `Gc` objects from inside
+the `drop` method a `Gc`. Alloy ensures that this rule is followed by checking
+for potential misuses of `drop` at compile-time in a pass known as _Finaliser
```
Comment (ltratt) [inline README.md:211]: s/in a pass known as/with/

File: README.md (hunk 5 above)
Comment (jacob-hughes) [reply README.md:246]: Fixed here: 29dcc937228ac01bb19a048a1b092fafd08d488a

Comment (ltratt) [reply README.md]: Not sure you've pushed that commit?

File: README.md [hunk 7]
```diff
@@ -199,16 +199,54 @@ that it uses existing drop methods as garbage collection finalisers, saving
...
+...
+2  |         *self.next;
+   |          ---------
+   |          |
+   |          caused by the expression here in `fn drop(&mut)` because
+   |          it uses another `Gc` type.
+   |
+   = help: `Gc` finalizers are unordered, so this field may have already been dropped. It is not safe to dereferenc
+e.
+```
```
[the remark this answers is not in the document]
Comment (jacob-hughes) [reply README.md:248]: Fixed here: 4450e64dbd345d430d2c1ca651cc6511d08b23b7

File: README.md (hunk 5 above)
Comment (jacob-hughes) [reply README.md]: I raced with github. Should work now

File: README.md (hunk 6 above)
Comment (jacob-hughes) [reply README.md]: Fixed here: 3855ba4824e745bca5ddb9e1ddc8b75455c3680f

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacob-hughes) [reply README.md:195]: Here 631c37a6fd02871ecc3f57304df65f8deca82ff0

**Comment by jacob-hughes** [comment]
I think I've addressed all your comments now :crossed_fingers:

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**